### ISSUE TITLE
Add UpdatableRecord.merge()

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#2961]
`UpdatableRecord.store()` executes an `INSERT` or `UPDATE` based on whether the primary key value has been previously loaded from the database. However, sometimes, people want to load `Record` values (including primary key) from some other DTO and "store" it in a similar fashion through an `INSERT .. ON DUPLICATE KEY UPDATE` statement, or through a `MERGE` statement.

This should be possible through `UpdatableRecord` API

Implementing this feature requires also:

- [ ] Additional `RecordListener` lifecycle events: #10042
- [ ] Correct handling of optimistic locking (might prove hard!)
- [ ] `DSLContext.batchMerge()`

---

See also these threads:
- https://groups.google.com/forum/#!topic/jooq-user/j6kk1CDJIjM
- https://groups.google.com/d/msg/jooq-user/NRjbXIrPEfc/zsSM5SykBAAJ
- #9687

**Comment by lukaseder** [comment]
The behaviour of `merge()` may seem a bit unintuitive at first. When loading a record, changing a single value, and then calling `merge()` on it, it is unlikely the statement will work as expected:

```java
var record = load();
record.setValue("a");
record.merge();
```

The above will create a statement similar to this one:

```sql
INSERT INTO t (value) VALUES ('a') ON DUPLICATE KEY UPDATE SET value = 'a'
```

It is effectively an *INSERT* statement of the *changed* values, ignoring the unchanged ones (e.g. the primary key). Users may have the insert or update semantics in mind, as in "store it back to the database". That's what the existing `store()` method is for.

`merge()` is best used when loading an *entire* record from the UI (e.g. from JSON), and then storing *all* of the data into the database without knowing whether the record already exists. So, the primary key would need to be included in the set of changed values.

**Comment by lukaseder** [comment]
This issue is implemented in jOOQ 3.14. Related work will be implemented in separate issues.
